server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -978,11 +978,16 @@ static List<SearchShardIterator> getRemoteShardsIteratorFromPointInTime(
         Map<String, OriginalIndices> remoteClusterIndices
     ) {
         final List<SearchShardIterator> remoteShardIterators = new ArrayList<>();
+        Map<ShardId, SearchContextIdForNode> searchContextIdForNodeMap = new HashMap<>(searchContextId.shards());
```

Comment:
Seems like we are just copying the map for the assertions sake?

Reply:
yes

Reply:
we could make it more complicated and do the copy only when assertions are enabled perhaps.

Reply:
I guess doesn't have to be that much more complicated? Why not just have a separate assertion function before the main loop that runs the same loop on a copy and asserts after the fact.

So basically something like:

```java
boolean assertStuff(searchContextId) {
        Map<ShardId, SearchContextIdForNode> searchContextIdForNodeMap = new HashMap<>.     (searchContextId.shards());
       for (Map.Entry<String, SearchShardsResponse> entry : searchShardsResponses.entrySet()) {
             for (SearchShardsGroup group : entry.getValue().getGroups()) {
                 final ShardId shardId = group.shardId();
                 final SearchContextIdForNode perNode = searchContextIdForNodeMap.remove(shardId);
              }
         }
       assert searchContextIdForNodeMap.values()
             .stream()
             .noneMatch(searchContextIdForNode -> searchContextIdForNode.getClusterAlias() != null)
       return true;
}

Then we don't have to mutate anything in the production code and have no performance impact :)?

Reply:
That makes sense, thanks, done in 52aac556aa905d73bafbb4e35f9c5e4b0a003d6b